token-swap/program/src/curve/stable.rs

```diff
@@ -35,21 +40,43 @@ pub struct StableCurve {
     pub amp: u64,
 }
 
+construct_uint! {
+    pub struct U256(4);
+}
+
+fn checked_u8_power(a: &U256, b: u8) -> Option<U256> {
+    let mut result = *a;
+    for _ in 1..b {
+        result = result.checked_mul(*a)?;
+    }
+    Some(result)
+}
+
+fn checked_u8_mul(a: &U256, b: u8) -> Option<U256> {
+    let mut result = *a;
+    for _ in 1..b {
+        result = result.checked_add(*a)?;
+    }
+    Some(result)
+}
+
+fn almost_equal(a: &U256, b: &U256) -> bool {
+    if a > b {
+        a - b <= U256::one()
+    } else {
+        b - a <= U256::one()
+    }
+}
+
 /// d = (leverage * sum_x + d_product * n_coins) * initial_d / ((leverage - 1) * initial_d + (n_coins + 1) * d_product)
-fn calculate_step(
-    initial_d: u128,
-    leverage: u128,
-    sum_x: u128,
-    d_product: u128,
-    n_coins: u128,
-) -> Option<u128> {
-    let leverage_mul = leverage.checked_mul(sum_x)?;
-    let d_p_mul = d_product.checked_mul(n_coins)?;
+fn calculate_step(initial_d: &U256, leverage: u64, sum_x: u128, d_product: &U256) -> Option<U256> {
+    let leverage_mul = U256::from(leverage).checked_mul(sum_x.into())?;
+    let d_p_mul = checked_u8_mul(d_product, N_COINS)?;
 
-    let l_val = leverage_mul.checked_add(d_p_mul)?.checked_mul(initial_d)?;
+    let l_val = leverage_mul.checked_add(d_p_mul)?.checked_mul(*initial_d)?;
 
-    let leverage_sub = leverage.checked_sub(1)?.checked_mul(initial_d)?;
-    let n_coins_sum = n_coins.checked_add(1)?.checked_mul(d_product)?;
+    let leverage_sub = initial_d.checked_mul((leverage - 1).into())?;
```

Comment:
This is pedantic, but it helps with security reviews later -- can you do `leverage.checked_sub(1)?`?